File: src/distance_warning/scripts/set_threshold_service.py
#!/usr/bin/env python3

import rclpy
from rclpy.node import Node
from distance_warning_interfaces.srv import SetThreshold
from rclpy.parameter import Parameter


class SetThresholdService(Node):
    def __init__(self):
        super().__init__('set_threshold_service')
        
        self.MIN_THRESHOLD = 0.1
        self.MAX_THRESHOLD = 1.5
        
        self.declare_parameter('threshold', 0.5)
        self.current_threshold = self.get_parameter('threshold').get_parameter_value().double_value
        
        self.service = self.create_service(
            SetThreshold,
            'set_threshold',
            self.handle_set_threshold
        )
        
        self.get_logger().info(f'Set Threshold Service initialized with threshold: {self.current_threshold:.2f} m')

    def handle_set_threshold(self, request, response):
        new_threshold = float(request.threshold)
        
        if new_threshold < self.MIN_THRESHOLD:
            new_threshold = self.MIN_THRESHOLD
            self.get_logger().warn(f'Requested threshold too low, setting to minimum: {self.MIN_THRESHOLD:.2f} m')
        elif new_threshold > self.MAX_THRESHOLD:
            new_threshold = self.MAX_THRESHOLD
            self.get_logger().warn(f'Requested threshold too high, setting to maximum: {self.MAX_THRESHOLD:.2f} m')
        
        param = Parameter('threshold', Parameter.Type.DOUBLE, new_threshold)
        self.set_parameters([param])
        self.current_threshold = new_threshold
        
        response.success = True
        
        self.get_logger().info(f'Threshold updated to: {self.current_threshold:.2f} m. Other nodes will get this value when they request it.')
        
        return response


def main(args=None):
    rclpy.init(args=args)
    node = SetThresholdService()
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass
    finally:
        node.destroy_node()
        rclpy.shutdown()


if __name__ == '__main__':
    main() 

File: src/distance_warning/src/set_threshold_service.cpp
#include "distance_warning/set_threshold_service.hpp"

SetThresholdService::SetThresholdService()
    : Node("set_threshold_service")
{
    this->declare_parameter<float>("threshold", 0.5f);

    m_service = this->create_service<SetThreshold>(
        "set_threshold",
        std::bind(&SetThresholdService::handle_set_threshold, this,
                  std::placeholders::_1, std::placeholders::_2));
    
    RCLCPP_INFO(this->get_logger(), "Other nodes can request threshold parameter from this service");
}

void SetThresholdService::handle_set_threshold(
    const std::shared_ptr<SetThreshold::Request> request,
    std::shared_ptr<SetThreshold::Response> response)
{
    float new_threshold = request->data ? -0.1f : 0.1f;
    new_threshold += this->get_parameter("threshold").as_double();
    
    if (new_threshold < MIN_THRESHOLD || new_threshold > MAX_THRESHOLD) {
        RCLCPP_WARN(this->get_logger(), "Threshold out of range: %.2f m", new_threshold);
        response->success = false;
        return;
    }

    this->set_parameter(rclcpp::Parameter("threshold", new_threshold));
    response->success = true;
    
    RCLCPP_INFO(this->get_logger(), "Threshold updated to: %.2f m. Other nodes will get this value when they request it.", new_threshold);
}

int main(int argc, char **argv)
{
    rclcpp::init(argc, argv);
    auto node = std::make_shared<SetThresholdService>();
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}
